ICE: builtin attribute "should_panic" not handled by `CheckAttrVisitor`

<!--
[31mICE[0m: Rustc ./C4C3F08962148F2308E9235B5B8F9E46B1ACDF573C166E52B2E6E7F0005AB3D7.rs '' 'error: internal compiler error: compiler/rustc_passes/src/check_attr.rs:285:33: builtin attribute "should_panic" not handled by `CheckAttrVisitor`', 'error: internal compiler error: compiler/rustc_passes/src/check_attr.rs:285:33: builtin attribute "should_panic" not handled by `CheckAttrVisitor`'
File: /tmp/im/C4C3F08962148F2308E9235B5B8F9E46B1ACDF573C166E52B2E6E7F0005AB3D7.rs
-->

auto-reduced (treereduce-rust):
````rust
struct S {
    #[should_panic::skip]
    field: u8,
}


````

original:
````rust
// Non-builtin attributes do not mess with field visibility resolution (issue #67006).

mod internal {
    struct S {
        #[rustfmt::skip]
        pub(in crate::internal) field: u32 // OK
    }

    struct Z(
        #[rustfmt::skip]
        pub(in crate::internal) u8 // OK
    );
}

struct S {
    #[should_panic::skip]
    pub(in nonexistent) field: u8 //~ ERROR failed to resolve
}

struct TokenStream(
    #[rustfmt::skip]
    pub(in nonexistent) u8 //~ ERROR failed to resolve
);

fn main() {}

````

Version information
````
rustc 1.82.0-nightly (64ebd39da 2024-08-03)
binary: rustc
commit-hash: 64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b
commit-date: 2024-08-03
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0433]: failed to resolve: use of undeclared crate or module `should_panic`
 --> /tmp/icemaker_global_tempdir.YjHCgx05egvK/rustc_testrunner_tmpdir_reporting.sPNqKpdsgpau/mvce.rs:2:7
  |
2 |     #[should_panic::skip]
  |       ^^^^^^^^^^^^ use of undeclared crate or module `should_panic`

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.YjHCgx05egvK/rustc_testrunner_tmpdir_reporting.sPNqKpdsgpau/mvce.rs:4:2
  |
4 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.YjHCgx05egvK/rustc_testrunner_tmpdir_reporting.sPNqKpdsgpau/mvce.rs`

error: internal compiler error: compiler/rustc_passes/src/check_attr.rs:285:33: builtin attribute "should_panic" not handled by `CheckAttrVisitor`
 --> /tmp/icemaker_global_tempdir.YjHCgx05egvK/rustc_testrunner_tmpdir_reporting.sPNqKpdsgpau/mvce.rs:2:5
  |
2 |     #[should_panic::skip]
  |     ^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at compiler/rustc_passes/src/check_attr.rs:285:33:
Box<dyn Any>
stack backtrace:
   0:     0x73cd3257515d - std::backtrace_rs::backtrace::libunwind::trace::h470c7ae87ba4025c
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x73cd3257515d - std::backtrace_rs::backtrace::trace_unsynchronized::h59a012957645db86
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x73cd3257515d - std::sys::backtrace::_print_fmt::he6eec48dee0520cd
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/sys/backtrace.rs:66:9
   3:     0x73cd3257515d - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hfdd25df09ce1fcfa
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/sys/backtrace.rs:39:26
   4:     0x73cd325c569b - core::fmt::rt::Argument::fmt::hb463cdd1073c5435
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/core/src/fmt/rt.rs:173:76
   5:     0x73cd325c569b - core::fmt::write::hd7c3b9720108f6ed
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/core/src/fmt/mod.rs:1178:21
   6:     0x73cd32568ec3 - std::io::Write::write_fmt::ha2df104fcce421c5
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/io/mod.rs:1823:15
   7:     0x73cd32577952 - std::sys::backtrace::BacktraceLock::print::hf06cfc93c67871f0
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/sys/backtrace.rs:42:9
   8:     0x73cd32577952 - std::panicking::default_hook::{{closure}}::hbb4ed713e2c9c643
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/panicking.rs:266:22
   9:     0x73cd325775be - std::panicking::default_hook::h7690590dddb77129
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/panicking.rs:293:9
  10:     0x73cd2eaa6d47 - std[5d818de65be8229c]::panicking::update_hook::<alloc[9a5ee3285ce50e9f]::boxed::Box<rustc_driver_impl[a2f6f36da10355a2]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x73cd32578342 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h79d7cc46a296b25c
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/alloc/src/boxed.rs:2164:9
  12:     0x73cd32578342 - std::panicking::rust_panic_with_hook::hbd0f660a3f2612ad
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/panicking.rs:805:13
  13:     0x73cd2eae2211 - std[5d818de65be8229c]::panicking::begin_panic::<rustc_errors[30d9fa35dd7a2581]::ExplicitBug>::{closure#0}
  14:     0x73cd2ead4f26 - std[5d818de65be8229c]::sys::backtrace::__rust_end_short_backtrace::<std[5d818de65be8229c]::panicking::begin_panic<rustc_errors[30d9fa35dd7a2581]::ExplicitBug>::{closure#0}, !>
  15:     0x73cd2ead0006 - std[5d818de65be8229c]::panicking::begin_panic::<rustc_errors[30d9fa35dd7a2581]::ExplicitBug>
  16:     0x73cd2eaeb481 - <rustc_errors[30d9fa35dd7a2581]::diagnostic::BugAbort as rustc_errors[30d9fa35dd7a2581]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  17:     0x73cd2f335e7d - <rustc_errors[30d9fa35dd7a2581]::DiagCtxtHandle>::span_bug::<rustc_span[e78ddc703a44fb87]::span_encoding::Span, alloc[9a5ee3285ce50e9f]::string::String>
  18:     0x73cd2f352818 - rustc_middle[60b1eb66b763f760]::util::bug::opt_span_bug_fmt::<rustc_span[e78ddc703a44fb87]::span_encoding::Span>::{closure#0}
  19:     0x73cd2f35284a - rustc_middle[60b1eb66b763f760]::ty::context::tls::with_opt::<rustc_middle[60b1eb66b763f760]::util::bug::opt_span_bug_fmt<rustc_span[e78ddc703a44fb87]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  20:     0x73cd2f34598b - rustc_middle[60b1eb66b763f760]::ty::context::tls::with_context_opt::<rustc_middle[60b1eb66b763f760]::ty::context::tls::with_opt<rustc_middle[60b1eb66b763f760]::util::bug::opt_span_bug_fmt<rustc_span[e78ddc703a44fb87]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  21:     0x73cd2f343b37 - rustc_middle[60b1eb66b763f760]::util::bug::span_bug_fmt::<rustc_span[e78ddc703a44fb87]::span_encoding::Span>
  22:     0x73cd302219d7 - <rustc_passes[82698ad2f6a923fe]::check_attr::CheckAttrVisitor>::check_attributes
  23:     0x73cd3021d944 - <rustc_passes[82698ad2f6a923fe]::check_attr::CheckAttrVisitor as rustc_hir[ea3fcf664761130]::intravisit::Visitor>::visit_item
  24:     0x73cd3021cdc0 - rustc_passes[82698ad2f6a923fe]::check_attr::check_mod_attrs
  25:     0x73cd3021cd3b - rustc_query_impl[952a2029dfc6d38d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[952a2029dfc6d38d]::query_impl::check_mod_attrs::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b1eb66b763f760]::query::erase::Erased<[u8; 0usize]>>
  26:     0x73cd30c13176 - rustc_query_system[def1941f6afd59f1]::query::plumbing::try_execute_query::<rustc_query_impl[952a2029dfc6d38d]::DynamicConfig<rustc_query_system[def1941f6afd59f1]::query::caches::DefaultCache<rustc_span[e78ddc703a44fb87]::def_id::LocalModDefId, rustc_middle[60b1eb66b763f760]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[952a2029dfc6d38d]::plumbing::QueryCtxt, false>
  27:     0x73cd30c12e99 - rustc_query_impl[952a2029dfc6d38d]::query_impl::check_mod_attrs::get_query_non_incr::__rust_end_short_backtrace
  28:     0x73cd30c12931 - <rustc_middle[60b1eb66b763f760]::hir::map::Map>::par_for_each_module::<rustc_interface[a7671a33eb98c4b3]::passes::run_required_analyses::{closure#0}::{closure#0}::{closure#1}::{closure#0}>::{closure#0}
  29:     0x73cd3059aefc - rustc_interface[a7671a33eb98c4b3]::passes::analysis
  30:     0x73cd3059abdb - rustc_query_impl[952a2029dfc6d38d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[952a2029dfc6d38d]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[60b1eb66b763f760]::query::erase::Erased<[u8; 1usize]>>
  31:     0x73cd30f7ac6e - rustc_query_system[def1941f6afd59f1]::query::plumbing::try_execute_query::<rustc_query_impl[952a2029dfc6d38d]::DynamicConfig<rustc_query_system[def1941f6afd59f1]::query::caches::SingleCache<rustc_middle[60b1eb66b763f760]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[952a2029dfc6d38d]::plumbing::QueryCtxt, false>
  32:     0x73cd30f7a9cf - rustc_query_impl[952a2029dfc6d38d]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  33:     0x73cd30e89be3 - rustc_interface[a7671a33eb98c4b3]::interface::run_compiler::<core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>, rustc_driver_impl[a2f6f36da10355a2]::run_compiler::{closure#0}>::{closure#1}
  34:     0x73cd30e27889 - std[5d818de65be8229c]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[a7671a33eb98c4b3]::util::run_in_thread_with_globals<rustc_interface[a7671a33eb98c4b3]::util::run_in_thread_pool_with_globals<rustc_interface[a7671a33eb98c4b3]::interface::run_compiler<core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>, rustc_driver_impl[a2f6f36da10355a2]::run_compiler::{closure#0}>::{closure#1}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>::{closure#0}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>
  35:     0x73cd30e27632 - <<std[5d818de65be8229c]::thread::Builder>::spawn_unchecked_<rustc_interface[a7671a33eb98c4b3]::util::run_in_thread_with_globals<rustc_interface[a7671a33eb98c4b3]::util::run_in_thread_pool_with_globals<rustc_interface[a7671a33eb98c4b3]::interface::run_compiler<core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>, rustc_driver_impl[a2f6f36da10355a2]::run_compiler::{closure#0}>::{closure#1}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>::{closure#0}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[dc2153a232c405cb]::result::Result<(), rustc_span[e78ddc703a44fb87]::ErrorGuaranteed>>::{closure#1} as core[dc2153a232c405cb]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x73cd3258205b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h4329d9c9b4cecba8
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/alloc/src/boxed.rs:2150:9
  37:     0x73cd3258205b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h2a2d295295b3a578
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/alloc/src/boxed.rs:2150:9
  38:     0x73cd3258205b - std::sys::pal::unix::thread::Thread::new::thread_start::hc3031069669580a0
                               at /rustc/64ebd39da5ec28caa3bd7cbb3f22f5949432fe2b/library/std/src/sys/pal/unix/thread.rs:105:17
  39:     0x73cd2b4a6ded - <unknown>
  40:     0x73cd2b52a0dc - <unknown>
  41:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.82.0-nightly (64ebd39da 2024-08-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_mod_attrs] checking attributes in top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0433, E0601.
For more information about an error, try `rustc --explain E0433`.

```

</p>
</details>

<!--
query stack:
#0 [check_mod_attrs] checking attributes in top-level module
#1 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_passes/src/check_attr.rs
+++ b/compiler/rustc_passes/src/check_attr.rs
@@ -116,130 +116,130 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
         let attrs = self.tcx.hir().attrs(hir_id);
         for attr in attrs {
             match attr.path().as_slice() {
-                [sym::diagnostic, sym::do_not_recommend] => {
+                [sym::diagnostic, sym::do_not_recommend, ..] => {
                     self.check_do_not_recommend(attr.span, hir_id, target)
                 }
-                [sym::diagnostic, sym::on_unimplemented] => {
+                [sym::diagnostic, sym::on_unimplemented, ..] => {
                     self.check_diagnostic_on_unimplemented(attr.span, hir_id, target)
                 }
-                [sym::inline] => self.check_inline(hir_id, attr, span, target),
-                [sym::coverage] => self.check_coverage(attr, span, target),
-                [sym::optimize] => self.check_optimize(hir_id, attr, target),
-                [sym::non_exhaustive] => self.check_non_exhaustive(hir_id, attr, span, target),
-                [sym::marker] => self.check_marker(hir_id, attr, span, target),
-                [sym::target_feature] => {
+                [sym::inline, ..] => self.check_inline(hir_id, attr, span, target),
+                [sym::coverage, ..] => self.check_coverage(attr, span, target),
+                [sym::optimize, ..] => self.check_optimize(hir_id, attr, target),
+                [sym::non_exhaustive, ..] => self.check_non_exhaustive(hir_id, attr, span, target),
+                [sym::marker, ..] => self.check_marker(hir_id, attr, span, target),
+                [sym::target_feature, ..] => {
                     self.check_target_feature(hir_id, attr, span, target, attrs)
                 }
-                [sym::thread_local] => self.check_thread_local(attr, span, target),
-                [sym::track_caller] => {
+                [sym::thread_local, ..] => self.check_thread_local(attr, span, target),
+                [sym::track_caller, ..] => {
                     self.check_track_caller(hir_id, attr.span, attrs, span, target)
                 }
-                [sym::doc] => self.check_doc_attrs(
+                [sym::doc, ..] => self.check_doc_attrs(
                     attr,
                     hir_id,
                     target,
                     &mut specified_inline,
                     &mut doc_aliases,
                 ),
-                [sym::no_link] => self.check_no_link(hir_id, attr, span, target),
-                [sym::export_name] => self.check_export_name(hir_id, attr, span, target),
-                [sym::rustc_layout_scalar_valid_range_start]
-                | [sym::rustc_layout_scalar_valid_range_end] => {
+                [sym::no_link, ..] => self.check_no_link(hir_id, attr, span, target),
+                [sym::export_name, ..] => self.check_export_name(hir_id, attr, span, target),
+                [sym::rustc_layout_scalar_valid_range_start, ..]
+                | [sym::rustc_layout_scalar_valid_range_end, ..] => {
                     self.check_rustc_layout_scalar_valid_range(attr, span, target)
                 }
-                [sym::allow_internal_unstable] => {
+                [sym::allow_internal_unstable, ..] => {
                     self.check_allow_internal_unstable(hir_id, attr, span, target, attrs)
                 }
-                [sym::debugger_visualizer] => self.check_debugger_visualizer(attr, target),
-                [sym::rustc_allow_const_fn_unstable] => {
+                [sym::debugger_visualizer, ..] => self.check_debugger_visualizer(attr, target),
+                [sym::rustc_allow_const_fn_unstable, ..] => {
                     self.check_rustc_allow_const_fn_unstable(hir_id, attr, span, target)
                 }
-                [sym::rustc_std_internal_symbol] => {
+                [sym::rustc_std_internal_symbol, ..] => {
                     self.check_rustc_std_internal_symbol(attr, span, target)
                 }
-                [sym::naked] => self.check_naked(hir_id, attr, span, target, attrs),
-                [sym::rustc_never_returns_null_ptr] => {
+                [sym::naked, ..] => self.check_naked(hir_id, attr, span, target, attrs),
+                [sym::rustc_never_returns_null_ptr, ..] => {
                     self.check_applied_to_fn_or_method(hir_id, attr, span, target)
                 }
-                [sym::rustc_legacy_const_generics] => {
+                [sym::rustc_legacy_const_generics, ..] => {
                     self.check_rustc_legacy_const_generics(hir_id, attr, span, target, item)
                 }
-                [sym::rustc_lint_query_instability] => {
+                [sym::rustc_lint_query_instability, ..] => {
                     self.check_rustc_lint_query_instability(hir_id, attr, span, target)
                 }
-                [sym::rustc_lint_diagnostics] => {
+                [sym::rustc_lint_diagnostics, ..] => {
                     self.check_rustc_lint_diagnostics(hir_id, attr, span, target)
                 }
-                [sym::rustc_lint_opt_ty] => self.check_rustc_lint_opt_ty(attr, span, target),
-                [sym::rustc_lint_opt_deny_field_access] => {
+                [sym::rustc_lint_opt_ty, ..] => self.check_rustc_lint_opt_ty(attr, span, target),
+                [sym::rustc_lint_opt_deny_field_access, ..] => {
                     self.check_rustc_lint_opt_deny_field_access(attr, span, target)
                 }
-                [sym::rustc_clean]
-                | [sym::rustc_dirty]
-                | [sym::rustc_if_this_changed]
-                | [sym::rustc_then_this_would_need] => self.check_rustc_dirty_clean(attr),
-                [sym::rustc_coinductive]
-                | [sym::rustc_must_implement_one_of]
-                | [sym::rustc_deny_explicit_impl]
-                | [sym::const_trait] => self.check_must_be_applied_to_trait(attr, span, target),
-                [sym::cmse_nonsecure_entry] => {
+                [sym::rustc_clean, ..]
+                | [sym::rustc_dirty, ..]
+                | [sym::rustc_if_this_changed, ..]
+                | [sym::rustc_then_this_would_need, ..] => self.check_rustc_dirty_clean(attr),
+                [sym::rustc_coinductive, ..]
+                | [sym::rustc_must_implement_one_of, ..]
+                | [sym::rustc_deny_explicit_impl, ..]
+                | [sym::const_trait, ..] => self.check_must_be_applied_to_trait(attr, span, target),
+                [sym::cmse_nonsecure_entry, ..] => {
                     self.check_cmse_nonsecure_entry(hir_id, attr, span, target)
                 }
-                [sym::collapse_debuginfo] => self.check_collapse_debuginfo(attr, span, target),
-                [sym::must_not_suspend] => self.check_must_not_suspend(attr, span, target),
-                [sym::must_use] => self.check_must_use(hir_id, attr, target),
-                [sym::rustc_pass_by_value] => self.check_pass_by_value(attr, span, target),
-                [sym::rustc_allow_incoherent_impl] => {
+                [sym::collapse_debuginfo, ..] => self.check_collapse_debuginfo(attr, span, target),
+                [sym::must_not_suspend, ..] => self.check_must_not_suspend(attr, span, target),
+                [sym::must_use, ..] => self.check_must_use(hir_id, attr, target),
+                [sym::rustc_pass_by_value, ..] => self.check_pass_by_value(attr, span, target),
+                [sym::rustc_allow_incoherent_impl, ..] => {
                     self.check_allow_incoherent_impl(attr, span, target)
                 }
-                [sym::rustc_has_incoherent_inherent_impls] => {
+                [sym::rustc_has_incoherent_inherent_impls, ..] => {
                     self.check_has_incoherent_inherent_impls(attr, span, target)
                 }
-                [sym::ffi_pure] => self.check_ffi_pure(attr.span, attrs, target),
-                [sym::ffi_const] => self.check_ffi_const(attr.span, target),
-                [sym::rustc_const_unstable]
-                | [sym::rustc_const_stable]
-                | [sym::unstable]
-                | [sym::stable]
-                | [sym::rustc_allowed_through_unstable_modules]
-                | [sym::rustc_promotable] => self.check_stability_promotable(attr, target),
-                [sym::link_ordinal] => self.check_link_ordinal(attr, span, target),
-                [sym::rustc_confusables] => self.check_confusables(attr, target),
-                [sym::rustc_safe_intrinsic] => {
+                [sym::ffi_pure, ..] => self.check_ffi_pure(attr.span, attrs, target),
+                [sym::ffi_const, ..] => self.check_ffi_const(attr.span, target),
+                [sym::rustc_const_unstable, ..]
+                | [sym::rustc_const_stable, ..]
+                | [sym::unstable, ..]
+                | [sym::stable, ..]
+                | [sym::rustc_allowed_through_unstable_modules, ..]
+                | [sym::rustc_promotable, ..] => self.check_stability_promotable(attr, target),
+                [sym::link_ordinal, ..] => self.check_link_ordinal(attr, span, target),
+                [sym::rustc_confusables, ..] => self.check_confusables(attr, target),
+                [sym::rustc_safe_intrinsic, ..] => {
                     self.check_rustc_safe_intrinsic(hir_id, attr, span, target)
                 }
-                [sym::cold] => self.check_cold(hir_id, attr, span, target),
-                [sym::link] => self.check_link(hir_id, attr, span, target),
-                [sym::link_name] => self.check_link_name(hir_id, attr, span, target),
-                [sym::link_section] => self.check_link_section(hir_id, attr, span, target),
-                [sym::no_mangle] => self.check_no_mangle(hir_id, attr, span, target),
-                [sym::deprecated] => self.check_deprecated(hir_id, attr, span, target),
-                [sym::macro_use] | [sym::macro_escape] => {
+                [sym::cold, ..] => self.check_cold(hir_id, attr, span, target),
+                [sym::link, ..] => self.check_link(hir_id, attr, span, target),
+                [sym::link_name, ..] => self.check_link_name(hir_id, attr, span, target),
+                [sym::link_section, ..] => self.check_link_section(hir_id, attr, span, target),
+                [sym::no_mangle, ..] => self.check_no_mangle(hir_id, attr, span, target),
+                [sym::deprecated, ..] => self.check_deprecated(hir_id, attr, span, target),
+                [sym::macro_use, ..] | [sym::macro_escape, ..] => {
                     self.check_macro_use(hir_id, attr, target)
                 }
-                [sym::path] => self.check_generic_attr(hir_id, attr, target, Target::Mod),
-                [sym::macro_export] => self.check_macro_export(hir_id, attr, target),
-                [sym::ignore] | [sym::should_panic] => {
+                [sym::path, ..] => self.check_generic_attr(hir_id, attr, target, Target::Mod),
+                [sym::macro_export, ..] => self.check_macro_export(hir_id, attr, target),
+                [sym::ignore, ..] | [sym::should_panic, ..] => {
                     self.check_generic_attr(hir_id, attr, target, Target::Fn)
                 }
-                [sym::automatically_derived] => {
+                [sym::automatically_derived, ..] => {
                     self.check_generic_attr(hir_id, attr, target, Target::Impl)
                 }
-                [sym::no_implicit_prelude] => {
+                [sym::no_implicit_prelude, ..] => {
                     self.check_generic_attr(hir_id, attr, target, Target::Mod)
                 }
-                [sym::rustc_object_lifetime_default] => self.check_object_lifetime_default(hir_id),
-                [sym::proc_macro] => {
+                [sym::rustc_object_lifetime_default, ..] => self.check_object_lifetime_default(hir_id),
+                [sym::proc_macro, ..] => {
                     self.check_proc_macro(hir_id, target, ProcMacroKind::FunctionLike)
                 }
-                [sym::proc_macro_attribute] => {
+                [sym::proc_macro_attribute, ..] => {
                     self.check_proc_macro(hir_id, target, ProcMacroKind::Attribute);
                 }
-                [sym::proc_macro_derive] => {
+                [sym::proc_macro_derive, ..] => {
                     self.check_generic_attr(hir_id, attr, target, Target::Fn);
                     self.check_proc_macro(hir_id, target, ProcMacroKind::Derive)
                 }
-                [sym::coroutine] => {
+                [sym::coroutine, ..] => {
                     self.check_coroutine(attr, target);
                 }
                 [
@@ -273,14 +273,16 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
                     | sym::default_lib_allocator
                     | sym::start
                     | sym::custom_mir,
+                    ..
                 ] => {}
                 [name, ..] => {
                     match BUILTIN_ATTRIBUTE_MAP.get(name) {
                         // checked below
                         Some(BuiltinAttribute { type_: AttributeType::CrateLevel, .. }) => {}
                         Some(_) => {
-                            // FIXME: differentiate between unstable and internal attributes just like we do with features instead
-                            // of just accepting `rustc_` attributes by name. That should allow trimming the above list, too.
+                            // FIXME: differentiate between unstable and internal attributes just
+                            // like we do with features instead of just accepting `rustc_`
+                            // attributes by name. That should allow trimming the above list, too.
                             if !name.as_str().starts_with("rustc_") {
                                 span_bug!(
                                     attr.span,

--- a/tests/ui/attributes/check-builtin-attr-ice.rs
+++ b/tests/ui/attributes/check-builtin-attr-ice.rs
@@ -1,0 +1,58 @@
+//! Regression test for #128622.
+//!
+//! PR #128581 introduced an assertion that all builtin attributes are actually checked via
+//! `CheckAttrVisitor` and aren't accidentally usable on completely unrelated HIR nodes.
+//! Unfortunately, the check had correctness problems.
+//!
+//! The match on attribute path segments looked like
+//!
+//! ```rs,ignore
+//! [sym::should_panic] => /* check is implemented */
+//! match BUILTIN_ATTRIBUTE_MAP.get(name) {
+//!     // checked below
+//!     Some(BuiltinAttribute { type_: AttributeType::CrateLevel, .. }) => {}
+//!     Some(_) => {
+//!         if !name.as_str().starts_with("rustc_") {
+//!             span_bug!(
+//!                 attr.span,
+//!                 "builtin attribute {name:?} not handled by `CheckAttrVisitor`"
+//!             )
+//!         }
+//!     }
+//!     None => (),
+//! }
+//! ```
+//!
+//! However, it failed to account for edge cases such as an attribute whose:
+//!
+//! 1. path segments *starts* with a builtin attribute such as `should_panic`
+//! 2. which does not start with `rustc_`, and
+//! 3. is also an `AttributeType::Normal` attribute upon registration with the builtin attribute map
+//!
+//! These conditions when all satisfied cause the span bug to be issued for e.g.
+//! `#[should_panic::skip]` because the `[sym::should_panic]` arm is not matched (since it's
+//! `[sym::should_panic, sym::skip]`).
+//!
+//! This test checks that the span bug is not fired for such cases.
+//!
+//! issue: rust-lang/rust#128622
+
+// Notably, `should_panic` is a `AttributeType::Normal` attribute that is checked separately.
+
+struct Foo {
+    #[should_panic::skip]
+    //~^ ERROR failed to resolve
+    pub field: u8,
+
+    #[should_panic::a::b::c]
+    //~^ ERROR failed to resolve
+    pub field2: u8,
+}
+
+fn foo() {}
+
+fn main() {
+    #[deny::skip]
+    //~^ ERROR failed to resolve
+    foo();
+}

--- a/tests/ui/attributes/check-builtin-attr-ice.stderr
+++ b/tests/ui/attributes/check-builtin-attr-ice.stderr
@@ -1,0 +1,21 @@
+error[E0433]: failed to resolve: use of undeclared crate or module `should_panic`
+  --> $DIR/check-builtin-attr-ice.rs:43:7
+   |
+LL |     #[should_panic::skip]
+   |       ^^^^^^^^^^^^ use of undeclared crate or module `should_panic`
+
+error[E0433]: failed to resolve: use of undeclared crate or module `should_panic`
+  --> $DIR/check-builtin-attr-ice.rs:47:7
+   |
+LL |     #[should_panic::a::b::c]
+   |       ^^^^^^^^^^^^ use of undeclared crate or module `should_panic`
+
+error[E0433]: failed to resolve: use of undeclared crate or module `deny`
+  --> $DIR/check-builtin-attr-ice.rs:55:7
+   |
+LL |     #[deny::skip]
+   |       ^^^^ use of undeclared crate or module `deny`
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0433`.

